presto-main/src/main/java/com/facebook/presto/server/protocol/QueryResourceUtil.java

```diff
@@ -274,4 +290,70 @@ private static int countStageAndAddGlobalUniqueNodes(List<TaskInfo> tasks, Set<S
         }
         return stageUniqueNodes.size();
     }
+
+    /**
+     * We follow the type signatures to track the data type and serializing nested data structure.
+     * Then deserialize it in the FixJsonDataUtils.
+     */
+    private static Iterable<List<Object>> prepareJsonData(List<Column> columns, Iterable<List<Object>> data)
+    {
+        if (data == null) {
+            return null;
+        }
+        requireNonNull(columns, "columns is null");
+        List<TypeSignature> signatures = columns.stream()
+                .map(column -> parseTypeSignature(column.getType()))
+                .collect(toList());
+        ImmutableList.Builder<List<Object>> rows = ImmutableList.builder();
+        for (List<Object> row : data) {
+            checkArgument(row.size() == columns.size(), "row/column size mismatch");
+            List<Object> newRow = new ArrayList<>();
+            for (int i = 0; i < row.size(); i++) {
+                newRow.add(parseToJson(signatures.get(i), row.get(i)));
+            }
+            rows.add(unmodifiableList(newRow)); // allow nulls in list
+        }
+        return rows.build();
+    }
+
+    public static Object parseToJson(TypeSignature signature, Object value)
+    {
+        if (value == null) {
+            return null;
+        }
+        if (signature.isDistinctType()) {
+            return parseToJson(signature.getDistinctTypeInfo().getBaseType(), value);
+        }
+        if (signature.getBase().equals(ARRAY)) {
+            List<Object> parsedValue = new ArrayList<>();
+            for (Object object : List.class.cast(value)) {
+                parsedValue.add(parseToJson(signature.getTypeParametersAsTypeSignatures().get(0), object));
+            }
+            return listJsonCodec(Object.class).toJson(parsedValue);
+        }
+        if (signature.getBase().equals(MAP)) {
+            TypeSignature keySignature = signature.getTypeParametersAsTypeSignatures().get(0);
+            TypeSignature valueSignature = signature.getTypeParametersAsTypeSignatures().get(1);
+            Map<Object, Object> parsedValue = new HashMap<>();
```

Comment:
Small nit, try to pass in a hint with the expected size of the map since it's known.

Reply:
passed size to it!